### PR TITLE
ISSUE #28: Retry on OPERATION_LIMITS_EXCEEDED

### PR DESCRIPTION
**Pull request by aledsage**
Note that to get the live tests working for me, I needed to do a few hacks (see https://github.com/cloudsoft/jclouds-vcloud-director/compare/1.9.x...live-test-hacks?expand=1). I have not included them in this PR - was it perhaps because of my system properties I use for testing? How do other people run the live tests?

### CODE REVIEW

**Comment by andreaturli** [comment]
Great work @aledsage

Minor suggestion about the overall implementation: have a look at https://github.com/jclouds/jclouds/blob/master/providers/digitalocean2/src/main/java/org/jclouds/digitalocean2/handlers/DigitalOcean2ErrorHandler.java I think following that recommended patternfor error handling on VCDErrorHandler could simplify a bit the rest of your patch.

File: src/main/java/org/jclouds/vcloud/director/v1_5/handlers/VcloudDirectorClientErrorRetryHandler.java [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+/**
+ * Handles Retryable responses with error codes in the 4xx range
+ */
+public class VcloudDirectorClientErrorRetryHandler extends BackoffLimitedRetryHandler {
+
+   private final VcloudDirectorUtils utils;
+   private final Set<String> retryableCodes;
+
+   @Inject
+   public VcloudDirectorClientErrorRetryHandler(VcloudDirectorUtils utils,
```
Comment (bostko) [inline src/main/java/org/jclouds/vcloud/director/v1_5/handlers/VcloudDirectorClientErrorRetryHandler.java]: I think it is recommended to use `@Inject` with protected or private constructor.

File: src/main/java/org/jclouds/vcloud/director/v1_5/config/VCloudDirectorHttpApiModule.java [hunk 2]
```diff
@@ -57,4 +79,9 @@ protected void installLocations() {
       bind(ImplicitLocationSupplier.class).to(OnlyLocationOrFirstZone.class).in(Scopes.SINGLETON);
    }
 
+   @Override
+   protected void bindRetryHandlers() {
+      bind(HttpRetryHandler.class).annotatedWith(ClientError.class).to(VcloudDirectorClientErrorRetryHandler.class);
+      bind(HttpRetryHandler.class).annotatedWith(ServerError.class).to(VcloudDirectorServerErrorRetryHandler.class);
```
Comment (andreaturli) [inline src/main/java/org/jclouds/vcloud/director/v1_5/config/VCloudDirectorHttpApiModule.java]: not sure why you want to bind ServerError when VcloudDirectorServerErrorRetryHandler has retryableServerCodes always empy.

Possibly you can remote entirely line 85?

**Comment by aledsage** [comment]
@andreaturli I'm very tempted to simplify the retry-handler code to be more like `DigitalOcean2ErrorHandler.java`, but on balance I think probably not. The reasons are:
- The other retry-handlers in jclouds all parse the message by calling a method on utils (e.g. the various aws retry-handlers, `AtmosClientErrorRetryHandler` and `AzureStorageClientErrorRetryHandler`.
- The digital ocean error handler example looks only at the contents of `respone.getStatusLine`, rather than the whole payload. Therefore doing a `message.contains()` feels a lot less risky than searching the entire response payload.
- Parsing the error means we will only retry if it is definitely an error xml object, with the right code.

Thoughts?

Note the DigitalOcean rate limit handler is entirely different impl, because the response explicitly says how long one should wait before retrying (so that pattern isn't really helpful): https://github.com/jclouds/jclouds/blob/master/providers/digitalocean2/src/main/java/org/jclouds/digitalocean2/handlers/RateLimitRetryHandler.java

For deleting `VcloudDirectorServerErrorRetryHandler`, I'll do that. I had wondered about leaving it in so that one could override the config that said what error codes to retry on - but given I didn't bother implementing it in `VcloudDirectorServerErrorRetryHandler`, then definitely better that we delete the code.

@bostko I'll change the injected constructor to be protected.

**Comment by aledsage** [comment]
@bostko @andreaturli I've incorporated those review comments. If you're happy for this to be merged, then let me know. I can squash the commits and merge (or you can!).

**Comment by bostko** [comment]
It may need a logger to notify when retry happened

**Comment by bostko** [comment]
@aledsage 
I added a logger locally on top of your latest code but I saw that jclouds is logging the client retries.
I tested it and I see messages like `Retry 1/6: delaying for 104 ms: server error`.
So the fix is working fine.

**Comment by aledsage** [comment]
Tested by @bostko - thanks. Merging now.

**Pull request merged**